stan/math/opencl/kernel_generator/load.hpp

```diff
@@ -0,0 +1,156 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_LOAD_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_LOAD_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation.hpp>
+#include <type_traits>
+#include <string>
+#include <utility>
+#include <set>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents an access to a \c matrix_cl in kernel generator expressions
+ * @tparam T \c matrix_cl
+ */
+template <typename T>
+class load__
+    : public operation<load__<T>, typename std::remove_reference_t<T>::type> {
+ public:
+  using ReturnScalar = typename std::remove_reference_t<T>::type;
+  using base = operation<load__<T>, ReturnScalar>;
+  using base::instance;
+  using base::var_name;
+  static_assert(std::is_base_of<matrix_cl<ReturnScalar>,
+                                typename std::remove_reference_t<T>>::value,
+                "load__: argument a must be a matrix_cl<T>!");
+  static_assert(
+      std::is_arithmetic<ReturnScalar>::value,
+      "load__: T in \"matrix_cl<T> a\" argument must be an arithmetic type!");
+
+  /**
+   * Constructor
+   * @param a \c matrix_cl
+   */
+  explicit load__(T&& a) : a_(std::forward<T>(a)) {}
```

Comment:
I think you need another template here if T can ever differ in cv or value category

Reply:
Here `T=decltype(a)`, except if a is rvalue reference of a type, than T is that type without reference. So it can never differ.

Reply:
Oh wuff, this is the constructor my bad